### PR TITLE
Sca rules

### PR DESCRIPTION
**Pull request by rzetelskik**

### CODE REVIEW

File: pkg/recommender/metrics/prometheus_provider.go [hunk 1]
```diff
@@ -60,40 +65,63 @@ func discover(ctx context.Context, c client.Client, metricsSelector map[string]s
...
+
+	resultMatrix := result.(model.Matrix)
+	if resultMatrix.Len() == 0 {
+		return false, errors.New("no results")
+	}
+
+	status := true
+	values := resultMatrix[0].Values //TODO check all results instead of just the first one???
+	for i := range values {
+		status = status && (values[i].Value != 0)
```
Comment (zimnx) [inline pkg/recommender/metrics/prometheus_provider.go:119]: you can early exit when status becomes false

File: pkg/recommender/metrics/provider.go [hunk 2]
```diff
@@ -2,15 +2,17 @@ package metrics
...
 import (
 	"context"
-	"github.com/prometheus/client_golang/api/prometheus/v1"
 	"github.com/scylladb/go-log"
+	"time"
 )
 
 type Provider interface {
-	FetchMetric(ctx context.Context, metric string, labels map[string]string) (float64, error)
+	Query(ctx context.Context, expression string) (bool, error)
```
Comment (zimnx) [inline pkg/recommender/metrics/provider.go:14]: It's a good practice to write short description of exposed functions in comment. Especially for exposed interfaces.

File: pkg/recommender/recommender.go [hunk 3]
```diff
@@ -36,97 +41,243 @@ func New(ctx context.Context, c client.Client, logger log.Logger, metricsSelecto
...
+		return nil, err
+	}
+
+	return sc, nil
+}
+
+func isScyllaClusterReady(sc *scyllav1.ScyllaCluster) bool {
+	for _, rack := range sc.Spec.Datacenter.Racks {
+		rackStatus, found := sc.Status.Racks[rack.Name]
+		if !found || rackStatus.Members != rackStatus.ReadyMembers {
```
Comment (zimnx) [inline pkg/recommender/recommender.go]: rackStatus.Members -> rack.Spec.Members

File: pkg/recommender/recommender.go [hunk 4]
```diff
@@ -36,97 +41,243 @@ func New(ctx context.Context, c client.Client, logger log.Logger, metricsSelecto
...
+
+	return true
+}
+
+func (r *recommender) getScyllaClusterRecommendations(ctx context.Context, sc *scyllav1.ScyllaCluster, scalingPolicy *v1alpha1.ScalingPolicy) (*v1alpha1.ScyllaClusterRecommendations, error) {
+	var datacenterRecommendations []v1alpha1.DatacenterRecommendations
+	for _, datacenterScalingPolicy := range scalingPolicy.Datacenters {
+		datacenter := sc.Spec.Datacenter
+		if datacenterScalingPolicy.Name != datacenter.Name {
+			return nil, errors.New(fmt.Sprintf("datacenter \"%s\" not found", datacenterScalingPolicy.Name))
```
Comment (zimnx) [inline pkg/recommender/recommender.go]: nit: errors.Errorf()

File: pkg/recommender/recommender.go [hunk 5]
```diff
@@ -36,97 +41,243 @@ func New(ctx context.Context, c client.Client, logger log.Logger, metricsSelecto
...
+			var min, max *int32 = nil, nil
+			if scalingPolicy.MemberPolicy != nil {
+				max = scalingPolicy.MemberPolicy.MaxAllowed
+				min = scalingPolicy.MemberPolicy.MinAllowed
+			}
+
+			members = calculateMembers(rack.Members, min, max, rule.ScalingFactor)
+		} else {
+			if rack.Resources.Requests == nil || rack.Resources.Requests.Cpu() == nil {
+				return nil, errors.New("cpu requests undefined")
```
Comment (zimnx) [inline pkg/recommender/recommender.go]: include info which rack doesn't have this info

**Review by zimnx — COMMENTED** [review]
Comments inline, please also add unit tests

File: pkg/recommender/recommender.go (hunk 5 above)
Comment (rzetelskik) [reply pkg/recommender/recommender.go]: the error is wrapped higher in the call stack (errors.Wrapf(err, "rack \"%s\"", rack.Name))

File: pkg/recommender/test/recommender_test.go [hunk 6]
```diff
@@ -0,0 +1,562 @@
+package test
```
Comment (zimnx) [inline pkg/recommender/test/recommender_test.go]: You don't have to create separate directory for test package. You can move these files to pkg/recommender and use `recommender_test` as package name

File: pkg/recommender/metrics/provider.go [hunk 7]
```diff
@@ -2,15 +2,31 @@ package metrics
...
+	// Perform a ranged query to the metrics provider.
+	// ctx - context.Context
+	// expression - an expression to be queried
+	// duration - time range
+	// step (optional) - minimal time range between each data point; optional, defaults to defaultStep
+	// Return value: unless an error has occurred, return a boolean value corresponding to the evaluated expression,
+	// i.e. whether the expression was true in each data point in the given time range.
+	RangedQuery(ctx context.Context, expression string, duration time.Duration, argStep *time.Duration) (bool, error)
+
+	SetApi(api v1.API)
```
Comment (zimnx) [inline pkg/recommender/metrics/provider.go]: You're binding generic interface with concrete implementation. Why would one have to implement accepting prometheus/v1 for different metrics backend?

File: pkg/recommender/metrics/test_util.go [hunk 8]
```diff
@@ -0,0 +1,21 @@
+package metrics
+
+import (
+	v1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	"github.com/scylladb/go-log"
+	"time"
+)
+
+func GetEmptyPrometheusProvider(logger log.Logger) Provider {
```
Comment (zimnx) [inline pkg/recommender/metrics/test_util.go]: do not return it as interface, but as a pointer to concrete implementation.

File: pkg/recommender/metrics/test_util.go [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+	return &prometheusProvider{
+		provider: provider{
+			logger:      logger,
+			defaultStep: time.Minute,
+		},
+		api: nil,
+	}
+}
+
+func (p *prometheusProvider) SetApi(api v1.API) {
```
Comment (zimnx) [inline pkg/recommender/metrics/test_util.go]: Instead requring this function in generic Provider interface, you can export it just for testing purposes.  
To have it available only in tests, you declare it as a part `metrics` package in `xxx_test.go` file (export_test.go for example). Then all `yyy_test.go` files within the same directory see this function as `metric.SetApi`, while non-test code cannot access it.

File: pkg/recommender/test/recommender_test.go (hunk 6 above)
Comment (zimnx) [reply pkg/recommender/test/recommender_test.go]: Usually package code and tests are shipped wihin the same directory. So for util package, you have util tests. Here you test utils as a part of recommender tests.

File: pkg/recommender/test/recommender_test.go [hunk 10]
```diff
@@ -0,0 +1,562 @@
...
+
+func TestPrometheusProviderQuery(t *testing.T) {
+
+	ctx := log.WithNewTraceID(context.Background())
+	atom := zap.NewAtomicLevelAt(zapcore.InfoLevel)
+	logger, _ := log.NewProduction(log.Config{
+		Level: atom,
+	})
+	r := recommender.GetEmptyRecommender(logger)
+	r.SetMetricsProviderFakeAPI(metrics.MockApi{})
```
Comment (zimnx) [inline pkg/recommender/test/recommender_test.go]: Mock should be injected as a constructor parameter. Constructor should accept interface type, and your test should be able to supply such.

File: pkg/recommender/test/recommender_test.go [hunk 11]
```diff
@@ -0,0 +1,562 @@
...
+
+	tests := []struct {
+		name           string
+		queryFun       func(string, time.Time) (model.Value, v1.Warnings, error)
+		queryExpr      string
+		expectedResult bool
+		errorExpected  bool
+	}{
+		{
+			name:           "Simple case",
```
Comment (zimnx) [inline pkg/recommender/test/recommender_test.go]: Just by reading test description I don't understand what is being tested, this should be improved.

File: pkg/recommender/recommender.go [hunk 12]
```diff
@@ -6,24 +6,31 @@ import (
...
+	"time"
 )
 
-type (
-	Recommender interface {
-		RunOnce(ctx context.Context) error
-	}
+type Recommender interface {
+	RunOnce(ctx context.Context) error
+	/*SetMetricsProviderFakeAPI(api metrics.MockApi)
```
Comment (zimnx) [inline pkg/recommender/recommender.go]: remove dead code

File: pkg/recommender/recommender.go [hunk 13]
```diff
@@ -36,97 +43,225 @@ func New(ctx context.Context, c client.Client, logger log.Logger, metricsSelecto
...
-					}
-				}
+	return nil
+}
 
-				if !found {
-					r.logger.Error(ctx, "rack not found", "rack", rackScalingPolicy.Name)
-					continue
-				}
+func (r *recommender) ExportedQueryOnMetricsProvider(ctx context.Context, expression string) (bool, error) {
```
Comment (zimnx) [inline pkg/recommender/recommender.go]: Changing application code just for testing purposes is bad practice. You should test packages as a blackbox, observing what is happening outside when exported function is called without having a need of calling private functions. You can mock/fake client.Client as well as metrics.Provider.

File: pkg/recommender/metrics/mock_prometheus_api.go [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+	"context"
+	"github.com/pkg/errors"
+	v1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	"github.com/prometheus/common/model"
+	"time"
+)
+
+type MockApi struct{}
+
+var Q = func(query string, ts time.Time) (model.Value, v1.Warnings, error) {
```
Comment (zimnx) [inline pkg/recommender/metrics/mock_prometheus_api.go]: Why this isn't a field of MockApi?

File: pkg/recommender/metrics/mock_prometheus_api.go [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+
+import (
+	"context"
+	"github.com/pkg/errors"
+	v1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	"github.com/prometheus/common/model"
+	"time"
+)
+
+type MockApi struct{}
```
Comment (zimnx) [inline pkg/recommender/metrics/mock_prometheus_api.go]: don't export it in application code. This should be moved to `prometheus_provider_mock_test.go` to be invisible outside of tests.

File: pkg/recommender/test/recommender_test.go [hunk 16]
```diff
@@ -0,0 +1,562 @@
...
+		scalingPolicy  *v1alpha1.RackScalingPolicy
+		expectedResult *v1alpha1.RackRecommendations
+		errorExpected  bool
+	}{
+		{
+			name: "Recommend scaling members",
+			rack: getRackSpec(rackName, baseMembers, baseCpu, baseCpu, memory, memory),
+			scalingPolicy: getScalingPolicy(rackName,
+				[]v1alpha1.ScalingRule{
+					*getScalingRule(ruleName, 1, queryWillReturnTrue, nil, nil, v1alpha1.ScalingModeHorizontal, 2.0),
```
Comment (zimnx) [inline pkg/recommender/test/recommender_test.go]: give a name to these magic numbers

**Review by zimnx — CHANGES_REQUESTED** [review]
go.sum change shouldn't be part of unit test commit
Tests should treat application as a blackbox, please refactor them to follow this appraoch of testing. 
Reorganization of code is needed.

File: pkg/recommender/test/recommender_test.go (hunk 16 above)
Comment (krzyssota) [reply pkg/recommender/test/recommender_test.go]: Ok. Added:
       const (
		rackName          = "rack_name"
		ruleName          = "rule_name"
		baseMembers       = 3
		baseCpu           = "5"
		higherCpu = "40"
		memory            = "1Gi"
		minAllowedMembers = 1
		maxAllowedMembers = 100
		priority1 = 1
		priority2 = 2
		priority3 = 3
		factor2 = 2.0
		factor4 = 4.0
		factor6 = 6.0
	)


	var (
		minAllowedCpu = resource.MustParse("1")
		maxAllowedCpu = resource.MustParse("100")
		duration5 = util.DurationPtr(5)
		duration10 = util.DurationPtr(10)
	)
	

and, if test results in scaling cpu, used it as follows:
	`corev1.ResourceCPU:    resource.MustParse(stringMulFloat64(baseCpu, factor4)),`

File: pkg/recommender/recommender.go (hunk 12 above)
Comment (krzyssota) [reply pkg/recommender/recommender.go]: Ok.

File: pkg/recommender/recommender.go (hunk 13 above)
Comment (krzyssota) [reply pkg/recommender/recommender.go:78]: "You should test packages as a blackbox, observing what is happening outside when exported function is called without having a need of calling private functions."
I tought unit test should cover single functions, some of which are private or are methods of interface private components (like Query of metricsProvider). Should I avoid exporting any function I do not have direct access to, i.e. writing only one big test for RunOnce (main function of recommender)?

File: pkg/recommender/metrics/test_util.go (hunk 8 above)
Comment (krzyssota) [reply pkg/recommender/metrics/test_util.go]: Solved it changing PrometheusProvider constructor.

File: pkg/recommender/metrics/provider.go (hunk 7 above)
Comment (krzyssota) [reply pkg/recommender/metrics/provider.go]: Got rid of it.

File: pkg/recommender/test/recommender_test.go (hunk 6 above)
Comment (krzyssota) [reply pkg/recommender/test/recommender_test.go]: Separated into recommender tests and prometheus provider tests.

File: pkg/recommender/metrics/test_util.go (hunk 9 above)
Comment (krzyssota) [reply pkg/recommender/metrics/test_util.go]: Solved it changing PrometheusProvider constructor.

File: pkg/recommender/test/recommender_test.go (hunk 10 above)
Comment (krzyssota) [reply pkg/recommender/test/recommender_test.go]: Mock is injected as an (interface typed) constructor parameter.

`func NewPrometheusProvider(api v1.API, logger log.Logger, defaultStep time.Duration) (Provider, error) {
	return &prometheusProvider{
		provider: provider{
			logger:      logger,
			defaultStep: defaultStep,
		},
		api: api,
	}, nil
}`

File: pkg/recommender/recommender.go (hunk 13 above)
Comment (krzyssota) [reply pkg/recommender/recommender.go]: Used fake client. Application code is not changed, calling only exported functions.

File: pkg/recommender/metrics/mock_prometheus_api.go (hunk 14 above)
Comment (krzyssota) [reply pkg/recommender/metrics/mock_prometheus_api.go]: Changed it.
`type mockApi struct {
	q  func(query string, ts time.Time) (model.Value, v1.Warnings, error)
	qr func(query string, r v1.Range) (model.Value, v1.Warnings, error)
}`

File: pkg/recommender/metrics/mock_prometheus_api.go (hunk 15 above)
Comment (krzyssota) [reply pkg/recommender/metrics/mock_prometheus_api.go]: Changed it.
`type mockApi struct {
	q  func(query string, ts time.Time) (model.Value, v1.Warnings, error)
	qr func(query string, r v1.Range) (model.Value, v1.Warnings, error)
}`

File: pkg/recommender/metrics/prometheus_provider.go [hunk 17]
```diff
@@ -18,38 +17,39 @@ import (
...
 	if len(svcList.Items) == 0 {
-		return nil, errors.Wrap(err, "no prometheus server found")
+		return nil, errors.New("no prometheus server found")
 	}
 
 	svc := svcList.Items[0]
 	addr := (&url.URL{
 		Scheme: "http",
-		Host:   fmt.Sprintf("%s.%s.svc.cluster.local", svc.Name, svc.Namespace),
+		Host:   fmt.Sprintf("%s.%s.svc.cluster.local:9090", svc.Name, svc.Namespace),
```
Comment (zimnx) [inline pkg/recommender/metrics/prometheus_provider.go]: 9090 could be a const

File: pkg/recommender/test_util.go [hunk 18]
```diff
@@ -0,0 +1,246 @@
+package recommender
```
Comment (zimnx) [inline pkg/recommender/test_util.go]: rename to util_test to make this code visible only in tests. If it's used in other packages, extract it to package.

File: pkg/recommender/metrics/mock_prometheus_api/mock_prometheus_api.go [hunk 19]
```diff
@@ -0,0 +1,99 @@
+package mock_prometheus_api
```
Comment (zimnx) [inline pkg/recommender/metrics/mock_prometheus_api/mock_prometheus_api.go]: just `mock` is enough

**Comment by zimnx** [comment]
Much better, thanks, left few nits.

File: pkg/recommender/metrics/prometheus_provider.go (hunk 17 above)
Comment (rzetelskik) [reply pkg/recommender/metrics/prometheus_provider.go]: added const svcPort

File: pkg/recommender/metrics/mock_prometheus_api/mock_prometheus_api.go (hunk 19 above)
Comment (rzetelskik) [reply pkg/recommender/metrics/mock_prometheus_api/mock_prometheus_api.go]: changed to mock

File: pkg/recommender/test_util.go (hunk 18 above)
Comment (rzetelskik) [reply pkg/recommender/test_util.go]: renamed to util_test.go

**Review by zimnx — APPROVED** [review]
LGTM, thanks

**Pull request merged**